### ISSUE TITLE
Some metrics are lost and I don't know how to debug it

### ISSUE THREAD

**Issue by lapo-luchini** [VictoriaMetrics/VictoriaMetrics#4972]
### Is your question request related to a specific component?

I don't know where the data is lost, I can't find any error and would like guidance on how to proceed on my debug.

### Describe the question in detail

I have a central single-node Victoria Metrics and a few decentralized `vmagent`s.
I have some values that are scraped correctly but cannot be found in the database with a query.

```mermaid
graph TD;
  vmagent1 --> vmauth;
  vmagent2 --> vmauth;
  vmagent3 --> vmauth;
  vmauth --> victoria-metrics;
```

This happens across different metrics in different scrapes, I will take one just as an example, but I have found no common rules among the "lost" metrics; they are all around (and rare enough, but still bothersome/alarming).

On http://vmagent1:8429/targets I can see:

| Endpoint | State | Labels | Debug relabeling | Scrapes | Errors | Last Scrape | Duration | Samples | Last error |
| -| -| - | - | - | - | - | - | - | - |
| http://vmagent1:9100/metrics (response) | UP | {instance="vmagent1:9100", job="node_exporter"} | target metrics | 887 | 0 | 12372ms ago | 4ms | 185

And the number of values seems to be correct:
```sh
% xh http://localhost:9100/metrics | egrep -cv '^#'
185
```

The server's `vmagent.log` reports no errors in writing and I get updated values for most of those 185 values. But… `count({datacenter="DC1",job="node_exporter"})` returns 122 (in internal `/vmui` and also in Grafana Explorer).

The `job` label is added by `vmagent` scraping rules:

```yaml
- job_name: 'node_exporter'
  static_configs:
  - targets:
    - vmagent1:9100
```

The `datacenter` label is forced by `vmauth`:

```yaml
users:

- username: "user1"
  password: "pass1"
  url_map:
  - src_paths: ["/api/v1/write"]
    url_prefix: "http://localhost:8428?extra_label=datacenter=DC1"
```

A missing metric is one I noticed because is used in Grafana to populate pop-ups in the upper-left corner, and that is `node_uname_info`, it just doesn't show up (and it is among the 185 scraped by connecting to node_exporter's :9100 output).

### Troubleshooting docs

- [x] General - https://docs.victoriametrics.com/Troubleshooting.html
- [x] vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
- [x] vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

**Comment by lapo-luchini** [comment]
I also tried "raw data" as suggested in troubleshooting, but the results are the same (no lines):
http://victoria:8428/api/v1/export?match[]=node_uname_info{datacenter="DC1"}&start=1694095940012

I'm currently running on `victoria-metrics-20230823-121336-tags-v1.93.1-0-gc8c20b7f7` (on FreeBSD) but this was happening also on 1.87 as far as I remember.

**Comment by lapo-luchini** [comment]
More data: this specific metric disappeared _exactly_ when I upgraded (both `victoria-metrics` and `vmauth`) from de9481208 to fc0116796 (1.91.2→1.93.0).

```
node_uname_info{datacenter="DC1"}
vm_app_version{job="victoria-metrics"}*2
vm_app_version{job="vmauth"}*3
```
<img width="616" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/420454/af6e970d-8b1b-4e33-9453-8fb170237c9b">

**Comment by lapo-luchini** [comment]
I don't really remember because this was one month ago, but in that "missing part" I guess that's when I downgraded (by mistake) from 1.91.2 to 1.87.x, which didn't start because of changed on-disk format, but then I upgraded to 1.93.0 and all went back ok. I wonder if/how this could impact future reception of data from that `uname` metrics.

**Comment by valyala** [comment]
It looks like data race issue has been introduced in [v1.93.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.93.1) in the commit a27c2f37731986f4bf6738404bb6388b1f42ffde , which addresses https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4247 . This issue results in corrupted labels for the collected samples by vmagent when the following conditions are met simultaneously:

- when `-remoteWrite.label` option is set
- when multiple `-remoteWrite.url` options are set.

The issue should be fixed in the commit 0bbc6a5b43209cb3b9c64bc2fbc9b33ef46b26df according to [these docs](https://docs.victoriametrics.com/vmagent.html#how-to-build-from-sources) . @lapo-luchini , could you build `vmagent` from this commit and verify whether the issue is gone now?

**Comment by lapo-luchini** [comment]
Sure, I'll build and upgrade that vmagent ASAP (probably on monday).

But I have two doubts:
1. I have only one remoteWrite URL set
2. I've seen the problem starting with 1.93.0, which should be before that issue

**Comment by lapo-luchini** [comment]
Built and running:
```sh
% vmagent --version
vmagent-20230909-135650-heads-master-0-g9bdcf483a
```
Unfortunately `count({datacenter="DC1",job="node_exporter"})` is still 122 instead of 185 (and that uname is still missing).

Any way I can help with debug?
I'm willing to build and try stuff.

PS: I only upgraded the remote `vmagent`, not the centralized `vmauth`+`victoria-metrics`, right?

**Comment by valyala** [comment]
> I have only one remoteWrite URL set
> I've seen the problem starting with 1.93.0, which should be before that issue
> Unfortunately count({datacenter="DC1",job="node_exporter"}) is still 122 instead of 185 (and that uname is still missing).

It looks like you hit another issue then. There are high chances the issue is in the `victoria-metrics` source code.

> Any way I can help with debug?

Could you start replicating the scraped data to a new `victoria-metrics` instance with the `vmagent` built from https://github.com/VictoriaMetrics/VictoriaMetrics/commit/0bbc6a5b43209cb3b9c64bc2fbc9b33ef46b26df ? E.g. you need to specify an additional `-remoteWrite.url` command-line flag to the `vmagent`, so it will replicate the scraped data to the new `victoria-metrics` instance. Then verify whether the new `victoria-metrics` instance v1.93.3 returns the correct number from `count({job="node_exporter"})` and whether it properly locates the `node_uname_info` metric. This should help determining whether the issue is reproduced on a fresh installation of `victoria-metrics` or it needs historical data.

If everything will be OK, then try adding `?extra_label=datacenter=DC1"` to the `-remoteWrite.url` of the new `victoria-metrics` and verifying again whether the new `victoria-metrics` returns the expected results. This should help whether the issue is related to the addition of extra labels via `extra_label` query arg during data ingestion.

**Comment by valyala** [comment]
FYI, the data race mentioned in [this commit](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4972#issuecomment-1712412380) has been fixed in `vmagent` [v1.93.4](https://docs.victoriametrics.com/CHANGELOG.html).

**Comment by lapo-luchini** [comment]
Upgraded `vmagent` to v1.93.4, changed nothing (of course).

Added a local `victoria-metrics` on vmagent1 server, added as a second remoteWrite, it does receive the metric.
Strangely `count({job="node_exporter"})` reports 191 instead of 185, but `node_uname_info` does show up correctly.

I then added `-remoteWrite.url=http://localhost:8428/api/v1/write?extra_label=datacenter=DC1` and it still works correctly (I also tried using a different value).

So it seems that reproducing the bug needs historical data, as I guess that `vmauth` role in the data is so small that it can be ignored. (?)

**Comment by lapo-luchini** [comment]
PS: "vmagent1" and "DC1" are placeholders to hide the actual names, but I guess the actual string content is not a problem (it is still fairly simple ASCII content, nothing strange). If it can help I have no problem in sharing a dump with devs, maybe not on a public issue tough.

**Comment by valyala** [comment]
@lapo-luchini , thanks for the provided information!

> Strangely count({job="node_exporter"}) reports 191 instead of 185

That's strange. Could you try exporting metrics for this target via [/federate API](https://docs.victoriametrics.com/#federation) and then comparing the exported metrics list with the metrics returned by the target and identifying the difference?

```
curl http://victoriametrics:8428/federate -d 'match[]={job="node_exporter"}'
```

BTW, it is better to filter by `instance` label instead of `job` label when searching for time series from a particular target, since `instance` label usually uniquely identifies the target, while multiple targets may have identical `job` label. See [these docs](https://prometheus.io/docs/concepts/jobs_instances/) for details.

> So it seems that reproducing the bug needs historical data, as I guess that vmauth role in the data is so small that it can be ignored. (?)

Correct - it is unlikely `vmauth` can introduce the bug, which leads to the loss of some metrics during data ingestion.

I suspect the issue may be related to the refactoring performed at [v1.92.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.92.0) for the following features:

- https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4563
- https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1401

I'm investigating these changes more thoroughly.

>  "vmagent1" and "DC1" are placeholders to hide the actual names, but I guess the actual string content is not a problem (it is still fairly simple ASCII content, nothing strange).

The exact contents of labels shouldn't matter.

**Comment by lapo-luchini** [comment]
> Could you try exporting metrics for this target via [/federate API](https://docs.victoriametrics.com/#federation)

Ah, ok, that `191` count is indeed right: those are the `185` from node_exporter port :9100 plus the `6` generated by scraping itself (which have the very same labels):

```
scrape_duration_seconds
scrape_samples_post_metric_relabeling
scrape_samples_scraped
scrape_series_added
scrape_timeout_seconds
up
```

> I suspect the issue may be related to the refactoring

Happy hunting!

If more tests are useful on my side, just let me know.

**Comment by valyala** [comment]
@lapo-luchini , could you run two queries via `curl` with the enabled [query tracing](https://docs.victoriametrics.com/#vmui) and responses of these queries here:

1. The query, which returns the requested metric: `curl http://victoriametrics:8428/api/v1/query -d trace=1 -d 'query=some_existing_metric{instance="..."}'
2. The query, which doesn't return the requested metric: `http://victoriametrics:8428/api/v1/query -d trace=1 -d 'query=node_uname_info{instance="..."}'

The difference between query tracing information in these responses can help identifying the issue.

I have a suspect that the issue can be related to a bug in the cache, which holds the list of time series ids matching the given [series filters](https://docs.victoriametrics.com/keyConcepts.html#filtering) (aka `indexdb/tagFiltersToMetricIDs` cache). In order to validate this suspect, you need to stop VictoriaMetrics, then to remove the `cache` directory inside the directory pointed by `-storageDataPath` command-line flag, and then start VictoriaMetrics again. Then verify whether VictoriaMetrics returns correct responses for both queries above.

**Comment by lapo-luchini** [comment]
> could you run two queries via `curl` with the enabled query tracing

I've done it and sent it to you directly on Telegram.

> In order to validate this suspect, you need to stop VictoriaMetrics, then to remove the `cache` directory

I've done that: the lines found are still only 122 (and the `uname` one doesn't appear).

Startup logs show this:

```
2023-09-13T09:23:31.238Z        info    …/VictoriaMetrics-1.93.4/app/vmstorage/main.go:122   successfully opened storage "/var/db/victoria" in 26.864 seconds; partsCount: 859; blocksCount: 436656893; rowsCount: 1570437878284; sizeBytes: 233919990921
2023-09-13T09:23:31.278Z        info    …/VictoriaMetrics-1.93.4/app/vmselect/promql/rollup_result_cache.go:114      loading rollupResult cache from "/var/db/victoria/cache/rollupResult"...
2023-09-13T09:23:31.280Z        info    …/VictoriaMetrics-1.93.4/app/vmselect/promql/rollup_result_cache.go:142      loaded rollupResult cache from "/var/db/victoria/cache/rollupResult" in 0.002 seconds; entriesCount: 0, sizeBytes: 0
```

**Comment by lapo-luchini** [comment]
Here are the two requested traces (with labels anonymized in the same way I did in the rest of the issue):
[traces-anon.tar.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/12596315/traces-anon.tar.gz)

**Comment by valyala** [comment]
The provided traces show that VictoriaMetrics cannot find TSID entry for the metricID associated with the `node_uname_info` time series, while it successfully finds the TSID entry for the metricID associated with the `node_time_seconds` time series.

- The `metricID` is a 64-bit number, which uniquely identifies every time series stored in VictoriaMetrics.
- The `TSID` is a data structure, which is used as a sorting and search key for data blocks stored on disk. It contains multiple other fields such as MetricGroupID, JobID and InstanceID additionally to MetricID. VictoriaMetrics sorts time series blocks by `TSID` in order to reduce the number of disk read operations needed to read time series data for the same metric name, the same `job` label and/or the same `instance` label.

VictoriaMetrics increments `vm_missing_tsids_for_metric_id_total` metric every time it cannot find `TSID` by the given `metricID`, since this is the expected case when the created `metricID->TSID` entry isn't available for search yet during a few seconds after registering new time series. It looks like in your case the `metricID->TSID` entry is permanently missing in the `indexdb`. This case is unexpected. This case may indicate to partial loss of `indexdb` data. The data may be lost there if some sub-directories under `<-storageDataPath>/indexdb/<indexDBGeneration>` directory were unexpectedly deleted. These sub-directories are also known as indexdb `parts`. VictoriaMetrics itself shouldn't delete these sub-directories unexpectedly - it deletes them only during background merge into bigger parts, and it makes sure that the source directories are deleted only after the resulting directory is created and completely saved to persistent storage. It is likely the subdirectories were deleted during accidental downgrade from to v1.91 to v1.87 mentioned [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4972#issuecomment-1711183528).

How to restore the deleted `metricID->TSID` entries? We can provide a custom VictoriaMetrics binary, which will re-create missing `metricID->TSID` entries during data ingestion. The needed `TSID` entry is located during data ingestion when VictoriaMetrics searches for `TSID` by the metric name plus all the labels for the ingested sample (this is known as `MetricName` in VictoriaMetrics source code).

You can run this VictoriaMetrics binary during a few scrape intervals in order to make sure `metricID->TSID` entries are re-created for all the [active time series](https://docs.victoriametrics.com/FAQ.html#what-is-an-active-time-series), and then return back to the official `v1.93.4` VictoriaMetrics binary. It is unsafe to run the custom binary for long period of time, since it may create duplicate `metricID->TSID` entries for newly ingested time series when these entries aren't visible for search yet.

There are chances that some other information is missing in the `indexdb` after the unexpected deletion of its parts. So other issues may arise in the future.

**Comment by valyala** [comment]
If you are OK with losing access to historical time series data for the missing `metricID->TSID` entries, then the way to go is to change some label in every ingested sample or to add a new label to all the ingested samples. This can be done by changing `extra_label` arg value passed to HTTP-based [data ingestion APIs](https://docs.victoriametrics.com/#how-to-import-time-series-data). In this case VictoriaMetrics will create new time series for all the ingested samples, and will generated and register new `TSID` with new `metricID->TSID` entries for these time series. After that newly ingested time series should be available during querying.

**Comment by lapo-luchini** [comment]
> It is likely the subdirectories were deleted during accidental downgrade from to v1.91 to v1.87 mentioned [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4972#issuecomment-1711183528).

Ohh, thanks for the analysis! This might be useful as a reference if anybody else had done that downgrade by mistake too, I hope.

> How to restore the deleted `metricID->TSID` entries? We can provide a custom VictoriaMetrics binary, which will re-create missing `metricID->TSID` entries during data ingestion.

That'd be nice! Just a few minute should be enough for all scrapes to arrive. If you have a small patch (or a fork / branch) I can compile that myself with no problems. (VM 1.93 is not yet officially on FreeBSD and [I'm trying to help with that](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=273257), so I'm compiliting it myself just to use it anyways)

**Comment by salarali** [comment]
I have similar issue where `metricID->TSID` entries are missing. vmstorage had panics because of this issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4873, which lead to issues where it deleted a lot of data? 

After restoring, I still have issues where a lot of metrics are missing. I checked using trace queries that the missing tsid's is the issue
I would be interested in using the binary in restoring the data.

**Comment by valyala** [comment]
>  I dont see the metric vm_missing_tsids_for_metric_id_total being incremented at all.

This metric is incremented only during queries, which select time series with missing `metricID->TSID` entries at `indexdb`.


@lapo-luchini and @salarali , you can build single-node VictoriaMetrics from the latest commit in the branch https://github.com/VictoriaMetrics/VictoriaMetrics/tree/issue-4972-add-missing-tsids (currently this is 641db141899e7b266b3d02c48ad58c12ae86cc8c ) according to [these docs](https://docs.victoriametrics.com/#how-to-build-from-sources) and try running it for a while - it should re-create missing `metricID->TSID` entries in the `indexdb`. It should log the following message per each missing entry:

```
Creating missing MetricID->TSID entry for MetricID=...
```

Note that it may create duplicate `MetricID->TSID` entries for newly registered time series. This should be OK.

**Comment by valyala** [comment]
FYI, the branch https://github.com/VictoriaMetrics/VictoriaMetrics/tree/issue-4972-add-missing-tsids-cluster contains the patch, which adds missing `metricID->TSID` entries for VictoriaMetrics cluster. `vmstorage` must be built from the latest commit in the branch according to [these docs](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#building-from-sources).

**Comment by valyala** [comment]
The following Docker image tags can be used for recovering missing `metricID->TSID` entries:

- Single-node VictoriaMetrics - `heads-issue-4972-add-missing-tsids-0-ge30e3387d`
- VictoriaMetrics cluster - `heads-issue-4972-add-missing-tsids-cluster-0-gbaaec4c3e`

**Comment by lapo-luchini** [comment]
The missing metric is back! :)
<img width="520" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/420454/89c38b77-bc25-4c8a-beda-bb1ba819a024">

**Comment by valyala** [comment]
@lapo-luchini , thanks for the update! Then closing this issue as resolved.

**Comment by anclrii** [comment]
I have a similar issue after upgrading my VictoriaMetrics cluster from `1.80.0` to `1.93.3`. I deployed the `heads-issue-4972-add-missing-tsids-cluster-0-gbaaec4c3e` tag for `vmstorage` . It starts `Creating missing MetricID->TSID entry` and after a few seconds it crashes as below:

```
2023-10-13T14:53:22.768Z        info    VictoriaMetrics/lib/storage/storage.go:2022     Creating missing MetricID->TSID entry for MetricID=1697208774613916197
2023-10-13T14:53:22.768Z        info    VictoriaMetrics/lib/storage/storage.go:2022     Creating missing MetricID->TSID entry for MetricID=1697208774613916198
2023-10-13T14:53:22.769Z        info    VictoriaMetrics/lib/storage/storage.go:2022     Creating missing MetricID->TSID entry for MetricID=1697208774613916199
2023-10-13T14:53:23.353Z        panic   VictoriaMetrics/lib/storage/index_db.go:2083    FATAL: cannot unmarshal metricName "\x00\x00\x00\x00\x00\x00\x00\x00,p)\x8ew1\xd9\xec\xd5s\xfd=\xdbg\x819\x17\x8d\xb1u\xed4\x96\xae": cannot unmarshal MetricGroup: cannot find the end of tag value
panic: FATAL: cannot unmarshal metricName "\x00\x00\x00\x00\x00\x00\x00\x00,p)\x8ew1\xd9\xec\xd5s\xfd=\xdbg\x819\x17\x8d\xb1u\xed4\x96\xae": cannot unmarshal MetricGroup: cannot find the end of tag value

goroutine 306 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logMessage({0x8e5782, 0x5}, {0xc0d624c000, 0xc4}, 0x2?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:309 +0xa91
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevelSkipframes(0x1, {0x8e5782, 0x5}, {0x8fa225?, 0x178db175ed3496ae?}, {0xc0d7205098?, 0x415c48?, 0xc0174dc360?})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:138 +0x199
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevel(...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:130
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.Panicf(...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:126
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).updateMetricIDsByMetricNameMatch(0xc01a1e48c0, 0x8fc1d4?, 0xc019a7d9e0, 0xc019a7d800, {0xc000502210, 0x1, 0x1})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2083 +0x5dc
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).getMetricIDsForDateAndFilters(0xc01a1e48c0, 0x0, 0x4cbb, 0xc026630240, 0x493e1)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2798 +0x1405
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).tryUpdatingMetricIDsForDateRange(0x60?, 0x58?, 0xc019a7d680, 0x155e?, {0x7feaebac95b0?, 0x7ff91cbbaf18?}, 0x60?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2550 +0x21a
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).updateMetricIDsForTagFilters(0xc01a1e48c0, 0x8fb758?, 0x1?, 0xc0d7205630?, {0x11?, 0x700?}, 0xc0d7205578?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2340 +0x33
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).searchMetricIDsInternal(0xc01a1e48c0?, 0x4?, {0xc0005021b8?, 0x1, 0x1}, {0x18b297d8305, 0x18b2986aac5}, 0x493e0)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2326 +0x416
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).searchMetricIDs(0xc01a1e48c0, 0x0?, {0xc0005021b8, 0x1, 0x1}, {0xc0d72057e8?, 0x0?}, 0x0?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2284 +0xad
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexDB).searchMetricIDs(0xc000140790, 0x8fc200?, {0xc0005021b0, 0x1, 0x1}, {0x18b297d8305, 0x18b2986aac5}, 0x493e0, 0x65295a02)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:1764 +0x41b
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*Search).Init(0xc001637000, 0xc0061a4640?, 0xc00018a9c0, {0xc0005021b0?, 0x1, 0x1}, {0x18b297d8305, 0x18b2986aac5}, 0x7ff8d5fc6650?, 0x65295a02)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/search.go:195 +0x2f5
github.com/VictoriaMetrics/VictoriaMetrics/app/vmstorage/servers.(*vmstorageAPI).InitSearch(0xc0004b27c8, 0xc0061a4600?, 0xc0061a4640, 0x0?)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmstorage/servers/vmselect.go:72 +0x17b
github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi.(*Server).processSearch(0xc000166640, 0xc0061a4600)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi/server.go:1033 +0x117
github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi.(*Server).processRPC(0xc0061a4600?, 0x7d9e8a927?, {0xc01a8271c0, 0x9})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi/server.go:561 +0x6b
github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi.(*Server).processRequest(0xc01360c0a0?, 0xc0061a4600)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi/server.go:513 +0x45f
github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi.(*Server).processConn(0xa00890?, 0xc0140bf020)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi/server.go:255 +0xbd
github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi.(*Server).run.func1()
        github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi/server.go:212 +0x228
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi.(*Server).run in goroutine 262
        github.com/VictoriaMetrics/VictoriaMetrics/lib/vmselectapi/server.go:180 +0x24a
```

Do I need to be on `1.93.4` before attempting this?

**Comment by valyala** [comment]
@anclrii , the given panic `suggestsFATAL: cannot unmarshal metricName` may occur after the corruption of the data stored on disk at `<-storageDataPath>/indexdb` folder, where `-storageDataPath` is the corresponding command-line flag value. In this case the only solution is to restore the data from [backups](https://docs.victoriametrics.com/vmbackup.html).

**Comment by lapo-luchini** [comment]
> try running it for a while - it should re-create missing `metricID->TSID` entries in the `indexdb`.

After a while I identified a single metric which is still missing, as this single one isn't very important historically (it's a blackbox metric to check a TLS certificate from expiration) I tried [deleting the series](https://docs.victoriametrics.com/guides/guide-delete-or-replace-metrics/) frmo DB, it was confirmed as empty, then after one minute new data was scraped and it came back… but still the single missing series was not showing up.

Is it possible to recover a single missing line (and, I guess, missing `metricID->TSID` entry) with API (even deleting data, as for that specific metric I don't care), or should I try merging that old branch with current version I'm using and have it runing for a few more minutes to recover this series mapping as well?
